stan/math/rev/fun/quad_form.hpp

```diff
@@ -93,6 +94,131 @@ class quad_form_vari : public vari {
 
   quad_form_vari_alloc<Ta, Ra, Ca, Tb, Rb, Cb>* impl_;
 };
+
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * Symmetry of the resulting matrix is not guaranteed due to numerical
+ * precision.
+ *
+ * @tparam Mat1 type of the first (square) matrix
+ * @tparam Mat2 type of the second matrix
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @param symmetric indicates whether the output should be made symmetric
+ * @return The quadratic form
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline auto quad_form_impl(const Mat1& A, const Mat2& B, bool symmetric) {
+  check_square("quad_form", "A", A);
+  check_multiplicable("quad_form", "A", A, "B", B);
+
+  using return_t
+      = promote_var_matrix_t<decltype(value_of(B).transpose().eval()
+                                      * value_of(A) * value_of(B).eval()),
+                             Mat1, Mat2>;
+
+  if (!is_constant<Mat1>::value && !is_constant<Mat2>::value) {
+    arena_t<promote_scalar_t<var, Mat1>> arena_A = A;
+    arena_t<promote_scalar_t<var, Mat2>> arena_B = B;
+
+    check_not_nan("multiply", "A", value_of(arena_A));
+    check_not_nan("multiply", "B", value_of(arena_B));
```

Comment:
Once we do `arena_t<promote_scalar_t<var, Mat1>> arena_A = A;` we know that `arena_A` has a var scalar so we can use `.val()` directly. I'm pretty sure `value_of()` is going to have to make a `holder()` type where calling `.val()` directly won't have to. Is there a reason you can't just use the version I wrote below that handles `var<mat>` and `mat<var>`?

https://github.com/stan-dev/math/compare/feature/quad_form-pullout-logic#diff-11ae2c18d3cbe94d5f48b9f855242f977ee3ec1d490c43723e09417f95c829f3R35

Reply:
The conclusion of this discussion that the `value_of` s here are fine and they don't create holders: https://github.com/stan-dev/math/pull/2164#discussion_r521897712 .

> Is there a reason you can't just use the version I wrote below that handles var<mat> and mat<var>?

I am basically using that code. The differences are:

1. `value_of` instead of worrying about the difference in `val` and `val_op`

2. We talked about at the last release to stop updating `mat<var>` code unless we have benchmarks showing it's faster, so I kept the old `mat<var>` stuff in place.

3. The code in that pull request returns a 1x1 matrix for the vector^T matrix vector case. This is not correct (see: https://mc-stan.org/docs/2_25/functions-reference/dot-products-and-specialized-products.html). I think there is a bug in the test framework that is letting this pass. I haven't made an issue for that yet.

Sound good?

Reply:
I made an issue for the 1x1 thing: https://github.com/stan-dev/math/issues/2198